homeassistant/components/uptimerobot/binary_sensor.py

```diff
@@ -24,17 +24,29 @@ async def async_setup_entry(
 ) -> None:
     """Set up the UptimeRobot binary_sensors."""
     coordinator = entry.runtime_data
-    async_add_entities(
-        UptimeRobotBinarySensor(
-            coordinator,
-            BinarySensorEntityDescription(
-                key=str(monitor.id),
-                device_class=BinarySensorDeviceClass.CONNECTIVITY,
-            ),
-            monitor=monitor,
-        )
-        for monitor in coordinator.data
-    )
+
+    known_devices: set[int] = set()
+
+    def _check_device() -> None:
+        current_devices = {monitor.id for monitor in coordinator.data}
+        new_devices = current_devices - known_devices
+        if new_devices:
+            known_devices.update(new_devices)
+            async_add_entities(
+                UptimeRobotBinarySensor(
+                    coordinator,
+                    BinarySensorEntityDescription(
+                        key=str(monitor.id),
+                        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+                    ),
+                    monitor=monitor,
+                )
+                for monitor in coordinator.data
+                if monitor.id in new_devices
+            )
+
+    _check_device()
+    entry.async_on_unload(coordinator.async_add_listener(_check_device))
```

Comment:
2 things with this approach:
1) Should this not be done in the coordinator instead of duplicating the implementation in each platform?
2) What about removals?

Reply:
that seems backwards to me, if in the coordinator it would handle both adding and removing, and apply to all platforms.

Reply:
Is requested to avoid reload: https://github.com/home-assistant/core/blob/dev/homeassistant/components/uptimerobot/quality_scale.yaml#L62

@joostlek can you step in and tell us which is the best approach ?

Reply:
I think there are more ways to solve it, it's a matter of context and taste. An integration like `tedee` keeps a `set` of all added devices, and if it encounters a new device in the data, it will call a separate callback listener that just adds the entities for that device. This can be beneficial for integrations with a lot of platforms.

The coordinator should not be tasked with creating the entities itself, that is part of the platforms.

Yes, reloading also works, but whenever possible, we strongly prefer the dynamic approach, which avoids entities going unavailable and just provides a smoother experience overall.